### DISCUSSION TITLE
Unify rasterized colors in photo of magazine comic strip

### DISCUSSION THREAD

**Discussion by forthrin** [ImageMagick/ImageMagick#7254] [Help]
### ImageMagick version

ImageMagick 7.1.1

### Operating system, version and so on

macOS 14

### Description

Consider the enclosed photo taken of a printed cartoon strip.

1. (Ignore the paper tear!)
2. Notice the rasterization on the pink, green and light/dark brown colors
3. Notice the intensity variation of the cyan background due to lighting.

Is there a fully automated command line that can accomplish the following (without the need for human handwork ahead)?

1. All basic colors (cyan, black, white, yellow, pink, green, light brown and dark brown) are unified into an 8 color palette.
2. Anti-aliasing colors between the basic colors and black are acceptable (or even wanted).
3. Either A), no resolution of the original lost, or preferably B), the image is vectorized, so that it can be up-scaled infinitely.

Background information:

- `potrace` can vectorize, but then all color is lost (and work must be done to whiten darker colors first)
- `pngquant` reduces colors intelligently, but often fails to unify large areas of similar colors in a satisfactory manner.
- ImageMagick can do stuff with "fuzz", but only used this on single colors earlier, not a set of colors.

Thanks in advance for kind help.

![Strip](https://github.com/ImageMagick/ImageMagick/assets/1167421/5f13161a-88b0-41d8-a5f7-c44d4723bf23)

**Comment by snibgo** [comment]
Here is a possible approach. Windows BAT syntax. For bash, change line-ending caret ^ to backslash \\, escape the parentheses \\( and \\), and don't double the percents %%.
```
magick ^
  comicStrip.jpg ^
  ( -clone 0 ^
    -fuzz 15%% -fill White +opaque sRGB(10%%,10%%,10%%) ^
    -write x1.png ^
    -write mpr:LINES ^
    -transparent White ^
    -write mpr:TRANS ^
    +delete ^
  ) ^
  mpr:LINES ^
  -alpha off -compose CopyOpacity -composite ^
  -blur 0x2 ^
  -write x2.png ^
  mpr:TRANS ^
  -compose Over -composite ^
  -write x3.png ^
  ( -clone 0 ^
    -fuzz 15%% -fill White +opaque sRGB(10%%,10%%,10%%) ^
    -write x4.png ^
    -write mpr:LINES ^
    -transparent White ^
    -write mpr:TRANS ^
    +delete ^
  ) ^
  mpr:LINES ^
  -alpha off -compose CopyOpacity -composite ^
  -blur 0x2 ^
  -write x5.png ^
  mpr:TRANS ^
  -compose Over -composite ^
  -write x6.png ^
  +dither -colors 10 ^
  -write x7.png ^
  NULL:
```
This clones the image, and turns pixels that are more than 15% from a 10% gray into white. So x1.png is black where the image is new black, otherwise it is white. mpr:TRANS is the same but black on transparent. We CopyOpacity to the input, and blur it. This smooths the cyan, pink etc colours. Then we copy over that the black lines, making x3.png.

The colours are not entirely smooth, so we do the same again, making x6.png. Finally, we reduce the image (in Lab colorspace) to 10 colours, making x7.png.

**Comment by forthrin** [comment]
Interesting! Not perfect, but it accomplishes some things.

So it seems you have to pick out parts of the image to work with individually before splicing them together?

But eg. the background is still banded. The cyan colors are very different from all the other colors in the photo, so isn't there a way to clearly separate the basic colors, and "flatten" everything within that color domain that clearly isn't one of the other colors?

Also, is there a vector tracer that supports colors?

How about separating the black, resizing the remaining stuff (the colors) to eg. 10% to make the colors blend, then resize it up again. Then vectorize the black and add it on top of the colors (the black outlines hopefully covering jagged edges of the resized colors).

Would something like that work?

**Comment by forthrin** [comment]
Does ImageMagick have functionality for recognizing areas, ie. be more flexible in considering adjacent colors to be the same than distant ones, eg. to understand that colors on the desk should be brown even if there are black-ish dithering there?

I thought a solution could be:

Convert colors to HSL, do weighted kMeans with k=8 on the HSL dimensions, and use the colors that emerge. That must surely separate the colors properly?

Or set a manual palette and force the converter to use that. Then the cyan shading must turn into a single cyan? When the alternative is yellow, etc. which is further away in the colorspace and will not be viable as candidates.

**Comment by snibgo** [comment]
Yes, both of those can be used. See `-kmeans` and `-remap` at https://imagemagick.org/script/command-line-options.php . And you can separate the image into areas of the same color, eg make an image that is white where the input is cyan, and black otherwise, then use a program such as potrace to vectorize each area.

**Comment by forthrin** [comment]
Tried `-kmeans`, but it refuses to merge all the cyan colors into one. Thus suspect the kmeans is not weighted, ie. it naively picks the most frequent pixel colors, of which there are several cyans.

`remap` with a hand-picked palette gives even worse results. (Granted, found no docs, so don't know what "1" and "255" means.)

```
# ImageMagick pixel enumeration: 8,1,255,RGB
0,0: (0,0,0)
1,0: (51,68,51)
2,0: (85,68,68)
3,0: (102,85,68)
4,0: (170,221,221)
5,0: (221,221,119)
6,0: (221,204,187)
7,0: (221,221,221)
```

Unless IM has an advanced "de-rasterize" filter or a "merge similar colors of continuous surfaces" filter, it seems necessary to hand-paint these areas prior to processing.
